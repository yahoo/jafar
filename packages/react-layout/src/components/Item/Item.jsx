/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import ItemView from '../ItemView';
import MobileItemView from '../MobileItemView';

export const layoutTypes = {
  TABS: 'tabs',
  SCROLL: 'scroll',
  MOBILE: 'mobile',
};
export default class Item extends React.Component {
  static propTypes = {
    title: PropTypes.string,
    layout: PropTypes.oneOf(Object.values(layoutTypes)),
    sections: PropTypes.array.isRequired,
    selected: PropTypes.object,
    mainActions: PropTypes.array,
    optionsActions: PropTypes.array,
    size: PropTypes.number,
  };

  constructor(props) {
    super(props);

    this.state = {
      selected: props.selected || {},
    };
  }

  render() {
    // filter section by exclude condition
    const filteredSections = filterSectionsByExcludeCondition(this.props.sections);

    if (filteredSections.length === 0) {
      return null;
    }

    const mainActions = this.props.mainActions ? prepareActions(this.props.mainActions) : undefined;
    const optionsActions = this.props.optionsActions ? prepareActions(this.props.optionsActions) : undefined;

    switch (this.props.layout) {
    case layoutTypes.MOBILE: {
      return (<MobileItemView title={this.props.title} sections={filteredSections} mainActions={mainActions}
        optionsActions={optionsActions} size={this.props.size} />);
    }
    default: {
      const selectedSectionId = this.state.selected.sectionId || filteredSections[0].id;

      // create the tabs menu items from the sections structure
      const tabs = this.props.layout ? {
        items: filteredSections.map(section => ({ value: section.id, label: section.title })),
        value: selectedSectionId,
        onChange: this.onChangeTab,
      } : undefined;

      this.mainActionsRefs = this.mainActionsRefs || {};
      (mainActions || []).forEach((action) => {
        if (action.popover) {
          if (!this.mainActionsRefs[action.label]) {
            this.mainActionsRefs[action.label] = React.createRef();
          }
          action.elementRef = this.mainActionsRefs[action.label];
          action.popover.targetRef = this.mainActionsRefs[action.label];
        }
      });

      this.sectionsRef = React.createRef();
      this.sectionsRefs = {};
      filteredSections.forEach((section) => {
        this.sectionsRefs[section.id] = React.createRef();
      });

      // prepare the current displayed sections and add ref to sections
      let sections = filteredSections.map(section => ({ ...section, ref: this.sectionsRefs[section.id] }));

      // if tabs layout - show only selected tab, without title for first level section
      if (this.props.layout === layoutTypes.TABS) {
        sections = [{ ...sections.find(section => section.id === selectedSectionId), title: undefined }];
      }

      // prepare footer
      const footer = this.props.mainActions ? { actions: mainActions } : undefined;

      // prepare options
      const options = this.props.optionsActions ? { actions: optionsActions } : undefined;

      return (<ItemView title={this.props.title} tabs={tabs} sections={sections} sectionsRef={this.sectionsRef}
        footer={footer} options={options} size={this.props.size} />);
    }
    }
  }

  static getDerivedStateFromProps(props, state) {
    if (props.selected && props.selected !== state.selected && props.selected !== state.prevSelected) {
      return ({
        needScroll: true,
        selected: props.selected,
        prevSelected: props.selected, // react doesn't provide prev props for compare here and in componentDidUpdate
      });
    }
    return null;
  }

  componentDidUpdate() {
    if (this.state.needScroll) {
      this.setState({
        needScroll: false,
      });

      // scroll to element
      if (this.state.selected.elementId) {
        const sections = this.sectionsRef.current;
        sections.scrollTo({
          top: sections.querySelector(`#${this.state.selected.elementId}`).offsetTop - sections.offsetTop - 10,
          left: 0,
          behavior: 'smooth',
        });
      }
    }
  }

  onChangeTab = (event, selectedSectionId) => {
    this.setState({
      selected: { sectionId: selectedSectionId, elementId: undefined },
    });

    // if scroll layout - scroll to section
    if (this.props.layout === layoutTypes.SCROLL) {
      this.sectionsRefs[selectedSectionId].current.scrollIntoView({ behavior: 'smooth' });
    }
  }
}

function filterSectionsByExcludeCondition(sections = []) {
  const newSections = sections.map(s => ({ ...s }));
  return newSections.filter((section) => {
    const renderSections = filterSectionsByExcludeCondition(section.sections);
    if (renderSections.length) {
      section.sections = renderSections;
    }
    return !(section.exclude && section.exclude()) || section.sections;
  });
}

function prepareActions(actions) {
  return actions
    .filter(action => !(action.exclude && action.exclude()))
    .map(action => Object.assign({}, action, {
      disabled: action.disable && action.disable(),
    }));
}
