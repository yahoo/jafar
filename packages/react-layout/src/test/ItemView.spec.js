/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import Item from '../components/ItemView/index';

describe('ItemView', () => {
  let props;

  beforeEach(() => {
    props = {
      title: 'Employee',
      sections: [{
        id: 'general',
      }],
      tabs: {
        items: [{
          value: 'general',
          label: 'General',
        }],
        value: undefined,
        onChange: jest.fn(),
      },
      footer: {
        actions: [],
      },
      options: {
        actions: [],
      },
    };
  });
  it('Should render - ok', () => {
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  it('Should render without title - ok', () => {
    props.title = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  it('Should render without tabs - ok', () => {
    props.tabs = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  it('Should render without footer - ok', () => {
    props.footer = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  it('Should render without options - ok', () => {
    props.options = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  it('Should render header options only - ok', () => {
    props.tabs = undefined;
    props.title = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  function getComponent(props) {
    return (<Item {...props} />);
  }
});
