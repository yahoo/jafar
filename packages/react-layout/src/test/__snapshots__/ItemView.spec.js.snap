// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemView Should render - ok 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      Employee
    </styled.div>
    <styled.div>
      <Options
        options={Array []}
      />
    </styled.div>
    <styled.div>
      <WithStyles(ForwardRef(Tabs))
        aria-label="Tabs"
        indicatorColor="primary"
        onChange={[MockFunction]}
        textColor="primary"
      >
        <WithStyles(ForwardRef(Tab))
          key="general"
          label="General"
          value="general"
        />
      </WithStyles(ForwardRef(Tabs))>
    </styled.div>
  </styled.div>
  <styled.div
    aria-label="Sections"
  >
    <ForwardRef
      id="general"
      key="general"
      size={4}
    />
  </styled.div>
  <Footer
    actions={Array []}
    aria-label="Footer"
  />
</styled.div>
`;

exports[`ItemView Should render header options only - ok 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <Options
        options={Array []}
      />
    </styled.div>
  </styled.div>
  <styled.div
    aria-label="Sections"
  >
    <ForwardRef
      id="general"
      key="general"
      size={4}
    />
  </styled.div>
  <Footer
    actions={Array []}
    aria-label="Footer"
  />
</styled.div>
`;

exports[`ItemView Should render without footer - ok 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      Employee
    </styled.div>
    <styled.div>
      <Options
        options={Array []}
      />
    </styled.div>
    <styled.div>
      <WithStyles(ForwardRef(Tabs))
        aria-label="Tabs"
        indicatorColor="primary"
        onChange={[MockFunction]}
        textColor="primary"
      >
        <WithStyles(ForwardRef(Tab))
          key="general"
          label="General"
          value="general"
        />
      </WithStyles(ForwardRef(Tabs))>
    </styled.div>
  </styled.div>
  <styled.div
    aria-label="Sections"
  >
    <ForwardRef
      id="general"
      key="general"
      size={4}
    />
  </styled.div>
</styled.div>
`;

exports[`ItemView Should render without options - ok 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      Employee
    </styled.div>
    <styled.div>
      <WithStyles(ForwardRef(Tabs))
        aria-label="Tabs"
        indicatorColor="primary"
        onChange={[MockFunction]}
        textColor="primary"
      >
        <WithStyles(ForwardRef(Tab))
          key="general"
          label="General"
          value="general"
        />
      </WithStyles(ForwardRef(Tabs))>
    </styled.div>
  </styled.div>
  <styled.div
    aria-label="Sections"
  >
    <ForwardRef
      id="general"
      key="general"
      size={4}
    />
  </styled.div>
  <Footer
    actions={Array []}
    aria-label="Footer"
  />
</styled.div>
`;

exports[`ItemView Should render without tabs - ok 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      Employee
    </styled.div>
    <styled.div>
      <Options
        options={Array []}
      />
    </styled.div>
  </styled.div>
  <styled.div
    aria-label="Sections"
  >
    <ForwardRef
      id="general"
      key="general"
      size={4}
    />
  </styled.div>
  <Footer
    actions={Array []}
    aria-label="Footer"
  />
</styled.div>
`;

exports[`ItemView Should render without title - ok 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <Options
        options={Array []}
      />
    </styled.div>
    <styled.div>
      <WithStyles(ForwardRef(Tabs))
        aria-label="Tabs"
        indicatorColor="primary"
        onChange={[MockFunction]}
        textColor="primary"
      >
        <WithStyles(ForwardRef(Tab))
          key="general"
          label="General"
          value="general"
        />
      </WithStyles(ForwardRef(Tabs))>
    </styled.div>
  </styled.div>
  <styled.div
    aria-label="Sections"
  >
    <ForwardRef
      id="general"
      key="general"
      size={4}
    />
  </styled.div>
  <Footer
    actions={Array []}
    aria-label="Footer"
  />
</styled.div>
`;
